.github/actions/csv-vulnerability-filter/src/processing.ts

```diff
@@ -0,0 +1,341 @@
+import * as core from '@actions/core';
+import * as fs from 'fs';
+import * as path from 'path';
+import Papa from 'papaparse';
+import { CsvRow, FilterResult, SeverityCounts, Thresholds } from './types';
+
+const AWS_INSPECTOR_HEADER_COLUMNS = [
+  "ID",
+  "Severity",
+  "CVSS",
+  "Installed Package",
+  "Fixed Package",
+  "Path",
+  "EPSS",
+  "Exploit Available",
+  "Exploit Last Seen",
+  "CWEs"
+];
+
+/**
+ * Reads and parses a CSV file, validating required columns.
+ * @param filePath Path to the CSV file.
+ * @param requiredColumns Array of column names expected to exist.
+ * @returns Array of row objects.
+ * @throws Error if file not found or required column is missing.
+ */
+export function parseCsv(filePath: string): CsvRow[] {
+  core.debug(`Attempting to read CSV file: ${filePath}`);
+  if (!fs.existsSync(filePath)) {
+    throw new Error(`CSV file not found at path: ${filePath}`);
+  }
+  const csvContent = fs.readFileSync(filePath, 'utf8');
+
+  // Check if file content is empty
+  if (!csvContent || csvContent.trim() === '') {
+    core.info(`CSV file ${filePath} is empty.`);
+    return [];
+  }
+
+  // Split content into lines, handling both \n and \r\n
+  const lines = csvContent.split(/\r?\n/);
+
+  // Check if there's at least a header line after removing the metadata line
+  if (lines.length <= 3) {
+    core.info(`CSV file ${filePath} contains only metadata line or is empty after splitting. No data to parse.`);
+    return [];
+  }
+
+  // Skip the first two lines of metadata. Header is the third line
+  const contentWithoutFirstLine = lines.slice(2).join('\n');
+
+  const parseResult = Papa.parse<CsvRow>(contentWithoutFirstLine, {
+    header: true,
+    skipEmptyLines: true,
+    dynamicTyping: false, // Keep all as strings initially
+  });
+
+  if (parseResult.errors.length > 0) {
+    core.warning(`CSV parsing encountered errors in ${filePath}:`);
+    parseResult.errors.slice(0, 5).forEach(err => core.warning(`  - ${err.message} (Row: ${err.row})`));
+    if (parseResult.errors.length > 5) {
+      core.warning(`  ... and ${parseResult.errors.length - 5} more errors.`);
+    }
+  }
+
+  if (!parseResult.data || parseResult.data.length === 0) {
+    core.info(`CSV file ${filePath} is empty or contains no data rows.`);
+    return [];
+  }
+
+  const headers = parseResult.meta.fields;
+  if (!headers) {
+    throw new Error(`Could not extract headers from CSV (after skipping first line): ${filePath}. Check file format.`);
+  }
+  core.info(`CSV Headers found in ${filePath} (after skipping first line): ${headers.join(', ')}`);
+
+  core.debug(`Successfully parsed ${parseResult.data.length} data rows from ${filePath}.`);
+  return parseResult.data;
+}
+
+function createFilteringKey(row: CsvRow): string | null {
+  return `${row['ID']}`;
+}
```

Comment:
This won't ever be `null`. If you want to allow it to return `undefined` instead of the string `'undefined'` you can remove the wrapping \`${...}\`
<img width="263" alt="Screenshot 2025-05-16 at 9 01 00 AM" src="https://github.com/user-attachments/assets/deec8f42-4335-4410-b6d5-7a0ad0c98b10" />

Reply:
Learned something!! changed.